cord/orm/dataset.py

```diff
@@ -15,65 +15,214 @@
+        """
+        This class has dict-style accessors for backwards compatibility.
+        Clients who are using this class for the first time are encouraged to use the property accessors and setters
+        instead of the underlying dictionary.
+        The mixed use of the `dict` style member functions and the property accessors and setters is discouraged.
+        """
+        super().__init__(
+            {
+                "data_hash": uid,
+                "data_title": title,
+                "data_type": data_type.to_upper_case_string(),
+                "created_at": created_at.strftime(DATETIME_STRING_FORMAT),
+            }
+        )
 
-class Dataset(base_orm.BaseORM):
-    """
-    A dataset is a collection of data rows.
+    @property
+    def uid(self) -> str:
+        return self.data["data_hash"]
 
-    ORM:
+    @uid.setter
+    def uid(self, value: str) -> None:
+        self.data["data_hash"] = value
 
-    title,
-    description,
-    dataset_type (Cord storage vs. AWS/GCP/Azure),
-    data_rows: [
-        {
-            data_hash (uid),
-            data_title,
-            data_type,
-        }
-    ]
+    @property
+    def title(self) -> str:
+        return self.data["data_title"]
 
-    """
+    @title.setter
+    def title(self, value: str) -> None:
+        self.data["data_title"] = value
 
-    DB_FIELDS = OrderedDict([("title", str), ("description", str), ("dataset_type", str), ("data_rows", (list, str))])
+    @property
+    def data_type(self) -> DataType:
+        return DataType.from_upper_case_string(self.data["data_type"])
 
-    NON_UPDATABLE_FIELDS = {
-        "dataset_type",
-    }
+    @data_type.setter
+    def data_type(self, value: DataType) -> None:
+        self.data["data_type"] = value.to_upper_case_string()
+
+    @property
+    def created_at(self) -> datetime:
+        return parser.parse(self.data["created_at"])
+
+    @created_at.setter
+    def created_at(self, value: datetime) -> None:
+        """Datetime will trim milliseconds for backwards compatibility."""
+        self.data["created_at"] = value.strftime(DATETIME_STRING_FORMAT)
+
+    @classmethod
+    def from_dict(cls, json_dict: Dict) -> DataRow:
+        data_type = DataType.from_upper_case_string(json_dict["data_type"])
+        if data_type is None:
+            raise TypeError(f"The DataRow constructor received an invalid data_type `{json_dict['data_type']}`")
+
+        return DataRow(
+            uid=json_dict["data_hash"],
+            title=json_dict["data_title"],
+            # The API server currently returns upper cased DataType strings.
+            data_type=data_type,
+            created_at=parser.parse(json_dict["created_at"]),
+        )
+
+    @classmethod
+    def from_dict_list(cls, json_list: List) -> List[DataRow]:
+        ret: List[DataRow] = list()
+        for json_dict in json_list:
+            ret.append(cls.from_dict(json_dict))
+        return ret
+
+
+class Dataset(UserDict, Formatter):
```

Comment:
@rad-cord the backwards compatibility was a bit more of a faff than I thought. I added the unit tests to make sure that I don't have a concentration error.

The only potential problem is if someone is using `Dataset.data` directly. Then they would break. Do you think we should prohibit access to it explicitly? The `data` member is how the `UserDict` is storing the fields.

Reply:
@rad-cord 🔔 this might be a deal breaker for our backwards compatibility effort.